drivers/counter/timer_esp32.c

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2020 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_timer
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <soc/rtc_cntl_reg.h>
+#include <soc/timer_group_reg.h>
+#include <driver/periph_ctrl.h>
+#include <soc/periph_defs.h>
+#include <hal/timer_types.h>
+#include <driver/timer.h>
+#include <hal/timer_hal.h>
+#include <soc.h>
+#include <string.h>
+#include <drivers/counter.h>
+#include <device.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(esp32_counter, CONFIG_COUNTER_LOG_LEVEL);
+
+#define INITIAL_COUNT (0x00000000ULL)
+
+#define INTR_SRC_0 ETS_TG0_T0_LEVEL_INTR_SOURCE
+#define INTR_SRC_1 ETS_TG0_T1_LEVEL_INTR_SOURCE
+#define INTR_SRC_2 ETS_TG1_T0_LEVEL_INTR_SOURCE
+#define INTR_SRC_3 ETS_TG1_T1_LEVEL_INTR_SOURCE
+
+#define INST_0_INDEX TIMER_0
+#define INST_1_INDEX TIMER_1
+#define INST_2_INDEX TIMER_0
+#define INST_3_INDEX TIMER_1
+
+#define INST_0_GROUP TIMER_GROUP_0
+#define INST_1_GROUP TIMER_GROUP_0
+#define INST_2_GROUP TIMER_GROUP_1
+#define INST_3_GROUP TIMER_GROUP_1
+
+#define TIMX p_timer_obj[TIMG(dev)][TIDX(dev)]
+#define DEV_CFG(dev) ((const struct counter_esp32_config *const)(dev)->config)
+#define DEV_DATA(dev) ((struct counter_esp32_data *)(dev)->data)
+#define TIMG(dev) (DEV_CFG(dev)->group)
+#define TIDX(dev) (DEV_CFG(dev)->idx)
+
+typedef void (*counter_irq_config_func_t)(const struct device *dev);
+
+struct timer_isr_func_t {
+	timer_isr_t fn;
+	void *args;
+	timer_isr_handle_t timer_isr_handle;
+	timer_group_t isr_timer_group;
+};
+
+struct counter_obj_t {
+	timer_hal_context_t hal;
+	struct timer_isr_func_t timer_isr_fun;
+};
+
+struct counter_esp32_config {
+	struct counter_config_info counter_info;
+	timer_config_t config;
+	timer_group_t group;
+	timer_idx_t idx;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	counter_irq_config_func_t irq_config_fn;
+};
+
+struct counter_esp32_data {
+	struct k_sem sem;
+	counter_alarm_callback_t callback;
+	uint32_t ticks;
+	void *user_data;
+};
+
+static struct counter_obj_t *p_timer_obj[TIMER_GROUP_MAX][TIMER_MAX] = { 0 };
+
+static inline void counter_esp32_sem_take(const struct device *dev)
+{
+	k_sem_take(&DEV_DATA(dev)->sem, K_FOREVER);
```

Comment:
You want a spinlock here, not a semaphore.  Semaphores are blocking IPC primitives.  They'll put the calling thread to sleep until the lock is released.  But I see below you're trying to take this same lock in an ISR, where clearly that isn't possible.  What will actually happen on contention is that the ISR will try to take the semaphore, fail to find a count to decrement, and then blow up with an assertion if you have that enabled (or otherwise fail spectacularly if you don't).

And it can't work even in principle: the thread that holds the lock might be the one you interrupted.